packages/dds/tree/src/test/feature-libraries/sequence-field/utils.ts

```diff
@@ -3,42 +3,105 @@
  * Licensed under the MIT License.
  */
 
+import { assert } from "@fluidframework/common-utils";
 import { ChangesetLocalId, IdAllocator, SequenceField as SF } from "../../../feature-libraries";
 import { Delta, TaggedChange, makeAnonChange, tagChange } from "../../../core";
 import { TestChange } from "../../testChange";
 import { assertFieldChangesEqual, deepFreeze, fakeRepair } from "../../utils";
-import { brand } from "../../../util";
+import { brand, fail } from "../../../util";
 import { TestChangeset } from "./testEdits";
 
 export function composeAnonChanges(changes: TestChangeset[]): TestChangeset {
-	const taggedChanges = changes.map(makeAnonChange);
-	return SF.sequenceFieldChangeRebaser.compose(
-		taggedChanges,
-		TestChange.compose,
-		continuingAllocator(taggedChanges),
-	);
+	return compose(changes.map(makeAnonChange));
 }
 
-export function rebaseTagged(
-	change: TaggedChange<TestChangeset>,
-	...base: TaggedChange<TestChangeset>[]
-): TaggedChange<TestChangeset> {
+export function composeNoVerify(changes: TaggedChange<TestChangeset>[]): TestChangeset {
+	return composeI(changes, (childChanges) => TestChange.compose(childChanges, false));
+}
+
+export function compose(changes: TaggedChange<TestChangeset>[]): TestChangeset {
+	return composeI(changes, TestChange.compose);
+}
+
+export function shallowCompose<T>(changes: TaggedChange<SF.Changeset<T>>[]): SF.Changeset<T> {
+	return composeI(changes, (children) => {
+		assert(children.length === 1, "Should only have one child to compose");
+		return children[0].change;
+	});
+}
+
+function composeI<T>(
+	changes: TaggedChange<SF.Changeset<T>>[],
+	composer: (childChanges: TaggedChange<T>[]) => T,
+): SF.Changeset<T> {
+	const moveEffects = SF.newCrossFieldTable();
+	const idAllocator = continuingAllocator(changes);
+	const composed = SF.compose(changes, composer, idAllocator, moveEffects);
+
+	if (moveEffects.isInvalidated) {
+		resetCrossFieldTable(moveEffects);
+		SF.amendCompose(composed, composer, idAllocator, moveEffects);
+		// assert(!moveEffects.isInvalidated, "Compose should not need more than one amend pass");
+	}
+	return composed;
+}
+
+export function rebase(change: TestChangeset, base: TaggedChange<TestChangeset>): TestChangeset {
 	deepFreeze(change);
 	deepFreeze(base);
 
+	const moveEffects = SF.newCrossFieldTable();
+	const idAllocator = idAllocatorFromMaxId(getMaxId(change, base.change));
+	let rebasedChange = SF.rebase(change, base, TestChange.rebase, idAllocator, moveEffects);
+	if (moveEffects.isInvalidated) {
+		moveEffects.reset();
+		rebasedChange = SF.amendRebase(rebasedChange, base, idAllocator, moveEffects);
+		// assert(!moveEffects.isInvalidated, "Rebase should not need more than one amend pass");
```

Comment:
Looks like we either need that assert back in or we need a loop.

Reply:
Currently this would loop infinitely due to problems with the invalidation logic. Although this is a fairly serious issue with this implementation, I would like to address it in a future PR.